src/runtime_src/hip/api/hip_memory.cpp

```diff
@@ -212,23 +212,49 @@ namespace xrt::core::hip
   {
-    throw_invalid_value_if(!src, "Invalid src pointer in hipMemCpyH2DAsync!");
+    throw_invalid_value_if(!src, std::string("src is nullptr in ") + __func__);
 
     auto hip_mem_info = memory_database::instance().get_hip_mem_from_addr(dst);
     auto hip_mem_dst = hip_mem_info.first;
     auto offset = hip_mem_info.second;
-    throw_invalid_value_if(!hip_mem_dst, "Invalid destination handle in hipMemCpyH2DAsync!");
-    throw_invalid_value_if(offset + size > hip_mem_dst->get_size(), "dst out of bound in hipMemCpyH2DAsync!");
+    throw_invalid_value_if(!hip_mem_dst, "invalid destination handle " + xrt_core::utils::to_hex(dst) + " in " + __func__);
+    throw_invalid_value_if(offset + size > hip_mem_dst->get_size(), "dst: " + xrt_core::utils::to_hex(dst) + " out of bound in " + __func__);
 
     auto hip_stream = get_stream(stream);
-    throw_invalid_value_if(!hip_stream, "Invalid stream handle in hipMemCpyH2DAsync!");
+    throw_invalid_value_if(!hip_stream, "invalid stream handle " + xrt_core::utils::to_hex(stream) + " in " + __func__);
    
     auto s_hdl = hip_stream.get();
     auto cmd_hdl = insert_in_map(command_cache,
-                                std::make_shared<copy_buffer>(hip_stream, XCL_BO_SYNC_BO_TO_DEVICE, hip_mem_dst, src, size, offset));
+                                std::make_shared<copy_buffer<void*>>(hip_stream, XCL_BO_SYNC_BO_TO_DEVICE, hip_mem_dst, src, size, offset));
     s_hdl->enqueue(command_cache.get(cmd_hdl));
   }
 
+  // fill data to dst async.
+  template<typename T> static void
+  hip_memset_async(void* dst, T value, size_t size, hipStream_t stream)
+  {
+    auto hip_mem_info = memory_database::instance().get_hip_mem_from_addr(dst);
+    auto hip_mem_dst = hip_mem_info.first;
+    auto offset = hip_mem_info.second;
+    assert(hip_mem_dst->get_type() != xrt::core::hip::memory_type::invalid);
+    throw_invalid_value_if(offset + size > hip_mem_dst->get_size(), "dst: " + xrt_core::utils::to_hex(dst) + " out of bound in " + __func__);
+
+    auto element_size = sizeof(T);
+
+    throw_invalid_value_if((element_size != 1 && element_size != 2 && element_size != 4), std::string("invalid element_type in ") + __func__);
+    throw_invalid_value_if(size % element_size != 0, "invalid size " + std::to_string(size) + " in " + __func__);
+
+    auto element_count = size / element_size;
+    auto vec = std::make_shared<std::vector<T>>(element_count, value);
+
+    auto hip_stream = get_stream(stream);
+    throw_invalid_value_if(!hip_stream, "invalid stream handle "  + xrt_core::utils::to_hex(stream) + " in " + __func__);
+
+    auto s_hdl = hip_stream.get();
+    auto cmd_hdl = insert_in_map(command_cache,
+                                std::make_shared<copy_buffer<std::shared_ptr<std::vector<T>>>>(hip_stream, XCL_BO_SYNC_BO_TO_DEVICE, hip_mem_dst, vec, size, offset));
+    s_hdl->enqueue(command_cache.get(cmd_hdl));
```

Comment:
I am not following what is going on here and there are no comments to help me.  The shared std::vector does not look right and a templated copy_buffer instantiated from a shared_ptr to std::vector does not seem right either.  Please carefully comment non obvious code.

Reply:
Ok, template use has been removed from class copy_buffer.